### DISCUSSION TITLE
bidirectional communication like in Arduino SoftwareSerial on one single line

### DISCUSSION THREAD

**Discussion by ralf33609** [arendst/Tasmota#12632] [Ideas]
Topic is slow bidirectional communication without RF or bluetooth or wifi 
In Arduino it is possible to init softserial with one single gpio: RX=TX=GPIOn. 

#include <SoftSerial>
#define RXPIN 10
#define TXPIN 10
SoftSerial myserial(RXPIN,TXPIN)

This would enable bidirectional communication over one wire with common Ground GND connected between both systems. 
Is this possible with Tasmotas softserial implementation? 
Even with slow baudrate 300baud ?
I would suggest for example to define one peripheral GPIO with "Serbr RX" or "Serial RX"  and an additional option command SETOPTIONxx 1 to define that GPIO_RX = GPIO_TX in the tasmotaserial.h  

IMHO: 1-wire protocol seems to be a bad solution because it neds a complicate slave implementation for one side of both systems.

**Comment by s-hadinger** [comment]
No it's not possible. Serial requires 2 wires by design, you can't short TX-TX from both sides. OneWire (I2C) is designed both electrically and protocol-wise to support bidirectional communications and handle smoothly collisions.

**Comment by lalo-uy** [comment]
The idea is to use open colector output. so both sides can pull down the
line and also receive on it if set hi.
Is up to protocols if there are collisions.

**Reply by sfromis** [reply]
Well, if you ensure that the tx signal from each device will not go back into rx, and that your protocol is strictly half duplex, what happens on the wires is outside of that the driver can react to. In principle, you could also let the signal be bridged back, and ignore the serial messages matching the ones you send.

**Comment by Jason2866** [comment]
Imho a lot of troubles (hard and software) to save one GPIO/wire.
What is the use case for?

**Reply by ralf33609** [reply]
the purpose is a frige. Outside the fridge i have a 5v power supply, a sensor input for energy(watt) and one actor for relay switching. inside the fridge I have as input a rotary encoder, a door switch atemperature sensor and as output a display. Outside to inside I have three cables. DC gnd,, vcc,, data.
Just for your information. I asked because I would like to generally solve this problem.
I read following discussion on seedstudio and maybe they dont use the standard arduino softserial library. At least Arno waldmann seems to have found a solution.
https://forum.seeedstudio.com/t/using-softwareserial-withonewire-rx-tx-pin-option-do-work/254885

**Reply by Jason2866** [reply]
Tasmota uses its own Software Serial implementation which is highly optimized for Tasmota needs. So do not expect that a corner use case will be merged if a PR is provided.

**Reply by ralf33609** [reply]
I did not really expect it. I thought myself that this kind of handling is difficult to handle in multitasking environment. So I just asked if a slight modifcation in softserial could handle this use case. before trying other solution. for example emulate a 1-wire slave eeprom on an avr to exchange data between tasmota and avr via 1-wire protocol.

**Comment by ralf33609** [comment]
I had a look at 
https://github.com/arduino/ArduinoCore-avr/blob/master/libraries/SoftwareSerial/src/SoftwareSerial.cpp
and I noticed that they dont check setRx pin and setTX pin in constructor not to be identical. So I assumed this implementation can handle it halfduplex. But I could not finally verify (lack of knowledge) in the code.

**Reply by Jason2866** [reply]
You have to look here https://github.com/arendst/Tasmota/tree/development/lib/default/TasmotaSerial-3.3.0

**Reply by ralf33609** [reply]
Aside of a pull request. As a first attempt to solve this for myself. Could detaching rxInterrupt while serial transmit and then attaching rxiterupt after transmission could be a Solution? (of course gpio mode handling has to be altered too. )

**Reply by s-hadinger** [reply]
Yes. That would be a required step to avoid conflict with Rx.

**Comment by ralf33609** [comment]
Well. then at first thanks to all. By the way really nice work on tasmota. 
I still cant really confirm that arduino library is capable of single wire softserial communication. 
I will test it and then try to adapt it to tasmotaserial. If I find a solution I will tell.
